### PR TITLE
CLOUD-579 - Add additional upgrade tests

### PR DESCRIPTION
**Pull request by tplavcic**
[CLOUD-579](https://jira.percona.com/browse/CLOUD-579)

Some notes:
- renames `smart-update` test group to `upgrade` and moves some tests there
- adds upgrade with haproxy and proxysql
- runs upgrade from last officially released PXC operator/cluster to the images that are currently tested (master or whatever PR)
- currently is missing `upgrade-proxy-protocol` test since it doesn't work - will be added later
- does not fix `smart-update` test, will be fixed in another PR, but don't want to add that here to not overcomplicate PR
- major version testing 5.7/8.0 and cluster wide tests work as other tests - test runs need to be started like that (select 5.7 images or `OPERATOR_NS` variable)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | failed |
| big-data | passed |
| scaling | passed |
| recreate | passed |
| haproxy | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| monitoring | failed |
| self-healing-advanced | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| affinity | failed |
| operator-self-healing | passed |
| one-pod | passed |
| auto-tuning | passed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| demand-backup-encrypted-with-tls | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| storage | failed |
| recreate | passed |
| scaling | passed |
| big-data | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| monitoring | failed |
| self-healing-advanced | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| affinity | failed |
| one-pod | passed |
| auto-tuning | passed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| demand-backup-encrypted-with-tls | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| storage | failed |
| haproxy | passed |
| recreate | passed |
| big-data | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| self-healing | passed |
| monitoring | failed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| self-healing-advanced | passed |
| affinity | failed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| demand-backup-encrypted-with-tls | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| scaling | passed |
| storage | failed |
| recreate | passed |
| haproxy | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| monitoring | failed |
| self-healing-advanced | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| affinity | failed |
| one-pod | passed |
| auto-tuning | passed |
| scheduled-backup | failed |
| proxysql-sidecar-res-limits | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| scaling | passed |
| haproxy | passed |
| storage | failed |
| big-data | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | passed |
| security-context | passed |
| limits | passed |
| monitoring | failed |
| restore-to-encrypted-cluster | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| affinity | failed |
| one-pod | passed |
| operator-self-healing | passed |
| auto-tuning | passed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| demand-backup-encrypted-with-tls | failed |

File: e2e-tests/upgrade-haproxy/run [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+    curl -s "https://raw.githubusercontent.com/percona/percona-xtradb-cluster-operator/${GIT_TAG}/deploy/${rbac}.yaml" > "${tmp_dir}/rbac_${GIT_TAG}.yaml"
+    $sed -i -e "s^namespace: .*^namespace: ${operator_namespace}^" "${tmp_dir}/rbac_${GIT_TAG}.yaml"
+    kubectl_bin apply -f "${tmp_dir}/rbac_${GIT_TAG}.yaml"
+}
+
+function deploy_operator_gh() {
+    desc 'start operator'
+    curl -s "https://raw.githubusercontent.com/percona/percona-xtradb-cluster-operator/${GIT_TAG}/deploy/crd.yaml" > "${tmp_dir}/crd_${GIT_TAG}.yaml"
+    kubectl_bin apply -f "${tmp_dir}/crd_${GIT_TAG}.yaml"
```
Comment (hors) [inline e2e-tests/upgrade-haproxy/run:121]: kubectl_bin apply -f  "https://raw.githubusercontent.com/percona/percona-xtradb-cluster-operator/${GIT_TAG}/deploy/crd.yaml"

File: e2e-tests/upgrade-proxysql/run [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        curl -s "https://raw.githubusercontent.com/percona/percona-xtradb-cluster-operator/${GIT_TAG}/deploy/cw-operator.yaml" > "${tmp_dir}/cw-operator_${GIT_TAG}.yaml"
+        $sed -i -e "s^image: .*^image: ${IMAGE}^" "${tmp_dir}/cw-operator_${GIT_TAG}.yaml"
+        kubectl_bin apply -f "${tmp_dir}/cw-operator_${GIT_TAG}.yaml" -n "${OPERATOR_NS}"
+    else
+        apply_rbac_gh rbac "${GIT_TAG}"
+        curl -s "https://raw.githubusercontent.com/percona/percona-xtradb-cluster-operator/${GIT_TAG}/deploy/operator.yaml" > "${tmp_dir}/operator_${GIT_TAG}.yaml"
+        $sed -i -e "s^image: .*^image: ${IMAGE}^" "${tmp_dir}/operator_${GIT_TAG}.yaml"
+        kubectl_bin apply -f "${tmp_dir}/operator_${GIT_TAG}.yaml"
+    fi
+
```
Comment (hors) [inline e2e-tests/upgrade-proxysql/run:134]:
```suggestion
    rbac_yaml='rbac'
    operator_yaml='operator.yaml'
    if [ -n "${OPERATOR_NS}" ]; then
        rbac_yaml='cw-rbac'
        operator_yaml='cw-operator.yaml'
    fi
    apply_rbac_gh $rbac_yaml "${GIT_TAG}"
    curl -s "https://raw.githubusercontent.com/percona/percona-xtradb-cluster-operator/${GIT_TAG}/deploy/$operator_yaml" > "${tmp_dir}/$operator_yaml_${GIT_TAG}.yaml"
   $sed -i -e "s^image: .*^image: ${IMAGE}^" "${tmp_dir}/$operator_yaml_${GIT_TAG}.yaml"
    kubectl_bin apply -f "${tmp_dir}/$operator_yaml_${GIT_TAG}.yaml" ${OPERATOR_NS:+-n $OPERATOR_NS}
```

File: Jenkinsfile [hunk 3]
```diff
@@ -256,11 +256,14 @@ pipeline {
...
             }
             parallel {
-                stage('E2E Smart update') {
+                stage('E2E Upgrade') {
                     steps {
-                        CreateCluster('smart-update')
-                        runTest('smart-update', 'smart-update')
-                        ShutdownCluster('smart-update')
+                        CreateCluster('upgrade')
+                        runTest('upgrade-haproxy', 'upgrade')
```
Comment (hors) [inline Jenkinsfile:262]: please update all pxc jenkins jobs too

**Review by hors — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 3 above)
Comment (tplavcic) [reply Jenkinsfile:262]: Added here: https://github.com/Percona-Lab/jenkins-pipelines/pull/623

File: e2e-tests/upgrade-proxysql/run (hunk 2 above)
Comment (tplavcic) [reply e2e-tests/upgrade-proxysql/run]: done, thanks!

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | failed |
| scaling | passed |
| haproxy | passed |
| recreate | failed |
| big-data | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | failed |
| security-context | failed |
| limits | passed |
| monitoring | failed |
| self-healing-advanced | failed |
| monitoring-2-0 | passed |
| affinity | failed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| restore-to-encrypted-cluster | failed |
| operator-self-healing | failed |

File: e2e-tests/upgrade-proxysql/run [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+function prepare_cr_yaml() {
+    local cr_yaml="$1"
+    local proxy="$2"
+
+    curl -s "https://raw.githubusercontent.com/percona/percona-xtradb-cluster-operator/${GIT_TAG}/deploy/cr.yaml" > "${cr_yaml}"
+    yq w -i "${cr_yaml}" "metadata.name" "${CLUSTER}"
+    yq w -i "${cr_yaml}" "spec.upgradeOptions.apply" "disabled"
+    yq w -i "${cr_yaml}" "spec.pxc.size" "${CLUSTER_SIZE}"
+    yq w -i "${cr_yaml}" "spec.proxysql.size" "${CLUSTER_SIZE}"
```
Comment (cap1984) [inline e2e-tests/upgrade-proxysql/run:162]: what do you think about passing curls output through yq pipes, not changing the file constantly?

Comment (tplavcic) [reply e2e-tests/upgrade-proxysql/run]: makes sense, updated! thanks!

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| scaling | passed |
| recreate | passed |
| big-data | passed |
| haproxy | passed |
| self-healing | failed |
| init-deploy | failed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | failed |
| demand-backup | failed |
| self-healing-advanced | failed |
| limits | failed |
| operator-self-healing | failed |
| monitoring | failed |
| scheduled-backup | failed |
| monitoring-2-0 | failed |
| affinity | failed |
| one-pod | failed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | failed |
| users | failed |
| tls-issue-self | failed |
| tls-issue-cert-manager | failed |
| tls-issue-cert-manager-ref | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| recreate | passed |
| big-data | failed |
| scaling | passed |
| haproxy | passed |
| self-healing | failed |
| init-deploy | failed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | failed |
| security-context | failed |
| self-healing-advanced | failed |
| limits | failed |
| demand-backup | failed |
| monitoring | failed |
| operator-self-healing | failed |
| scheduled-backup | failed |
| monitoring-2-0 | failed |
| affinity | failed |
| demand-backup-encrypted-with-tls | failed |
| one-pod | failed |
| auto-tuning | failed |
| proxysql-sidecar-res-limits | failed |
| users | failed |
| tls-issue-self | failed |
| tls-issue-cert-manager | failed |
| tls-issue-cert-manager-ref | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| storage | passed |
| big-data | passed |
| scaling | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| security-context | failed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| operator-self-healing | failed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| storage | passed |
| self-healing | passed |
| big-data | passed |
| self-healing-advanced | passed |
| operator-self-healing | failed |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| upgrade-haproxy | failed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | failed |
| upgrade-proxysql | failed |
| smart-update | failed |
| tls-issue-self | passed |
| tls-issue-cert-manager | failed |
| upgrade-consistency | failed |
| tls-issue-cert-manager-ref | failed |
| scheduled-backup | failed |
| demand-backup-encrypted-with-tls | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| haproxy | passed |
| big-data | passed |
| recreate | passed |
| scaling | passed |
| upgrade-haproxy | failed |
| storage | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| security-context | failed |
| restore-to-encrypted-cluster | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| demand-backup | failed |
| monitoring-2-0 | passed |
| operator-self-healing | failed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| upgrade-proxysql | passed |
| smart-update | failed |
| upgrade-consistency | failed |
| scheduled-backup | failed |
| tls-issue-self | failed |
| demand-backup-encrypted-with-tls | failed |
| tls-issue-cert-manager | failed |
| tls-issue-cert-manager-ref | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| storage | passed |
| recreate | passed |
| upgrade-haproxy | failed |
| big-data | passed |
| haproxy | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | failed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| demand-backup | failed |
| monitoring-2-0 | passed |
| operator-self-healing | failed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| upgrade-proxysql | passed |
| smart-update | failed |
| upgrade-consistency | failed |
| tls-issue-self | failed |
| scheduled-backup | failed |
| demand-backup-encrypted-with-tls | failed |
| tls-issue-cert-manager | failed |
| tls-issue-cert-manager-ref | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| scaling | passed |
| recreate | passed |
| storage | passed |
| upgrade-haproxy | passed |
| haproxy | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| self-healing | passed |
| security-context | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| smart-update | failed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| operator-self-healing | failed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| upgrade-consistency | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| scheduled-backup | passed |
| demand-backup-encrypted-with-tls | passed |

File: e2e-tests/upgrade-haproxy/run [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+    wait_cluster_consistency "${CLUSTER}" "${CLUSTER_SIZE}"
+
+    wait_for_running "${CLUSTER}-pxc" "${CLUSTER_SIZE}"
+
+    for i in $(seq 0 $((CLUSTER_SIZE-1))); do
+        compare_mysql_cmd "select-1" "SELECT * from myApp.myApp;" "-h ${CLUSTER}-pxc-${i}.${CLUSTER}-pxc -uroot -proot_password"
+    done
+}
+
+function compare_generation() {
```
Comment (hors) [inline e2e-tests/upgrade-haproxy/run:61]: we use the same function in two different tests, we need to move it to functions

File: e2e-tests/upgrade-haproxy/run [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    for container in "${containers[@]}"; do
+        current_generation="$(kubectl_bin get statefulset "${CLUSTER}-${container}" -o jsonpath='{.metadata.generation}')"
+        if [[ "${generation}" != "${current_generation}" ]]; then
+            echo "Generation for resource ${container} is: ${current_generation}, but should be: ${generation}"
+            exit 1
+        fi
+    done
+}
+
+function check_applied_images() {
```
Comment (hors) [inline e2e-tests/upgrade-haproxy/run:80]: can we move it to functions  file?

Comment (hors) [reply e2e-tests/upgrade-haproxy/run:80]: Please re-check all functions and move all needed (which are used more than in one place) stuff to functions file.

**Review by hors — CHANGES_REQUESTED** [review]

File: e2e-tests/upgrade-haproxy/run (hunk 5 above)
Comment (tplavcic) [reply e2e-tests/upgrade-haproxy/run]: done

File: e2e-tests/upgrade-haproxy/run (hunk 6 above)
Comment (tplavcic) [reply e2e-tests/upgrade-haproxy/run]: done

**Review by cap1984 — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| upgrade-haproxy | passed |
| recreate | passed |
| big-data | passed |
| storage | passed |
| scaling | passed |
| haproxy | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| upgrade-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| limits | passed |
| self-healing | passed |
| smart-update | failed |
| monitoring | passed |
| demand-backup | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| self-healing-advanced | passed |
| auto-tuning | passed |
| operator-self-healing | failed |
| upgrade-consistency | passed |
| scheduled-backup | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| demand-backup-encrypted-with-tls | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| upgrade-haproxy | passed |
| upgrade-proxysql | passed |
| smart-update | failed |
| big-data | passed |
| recreate | passed |
| storage | passed |
| haproxy | passed |
| restore-to-encrypted-cluster | passed |
| self-healing | passed |
| init-deploy | passed |
| demand-backup | passed |
| self-healing-advanced | passed |
| limits | passed |
| operator-self-healing | failed |
| scheduled-backup | passed |
| monitoring | passed |
| demand-backup-encrypted-with-tls | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | failed |
| upgrade-consistency | passed |

**Review by hors — APPROVED** [review]

**Pull request merged**